docs/architecture/adr-038-state-listening.md

```diff
@@ -0,0 +1,607 @@
+	types.AssertValidKey(key)
+	s.parent.Set(key, value)
+	s.onWrite(key, value)
+}
+
+// Delete implements the KVStore interface. It traces a write operation and
+// delegates the Delete call to the parent KVStore.
+func (s *Store) Delete(key []byte) {
+	s.parent.Delete(key)
+	s.onWrite(key, nil)
+}
+
+// onWrite writes a KVStore operation to all of the WriteListeners
+func (s *Store) onWrite(key, value []byte) {
+	for _, l := range s.listeners {
+		l.OnWrite(s.parentStoreKey, key, value)
+	}
+}
+```
+
+### MultiStore interface updates
+
+We will update the `MultiStore` interface to allow us to wrap a set of listeners around a specific `KVStore`.
+Additionally, we will update the `CacheWrap` and `CacheWrapper` interfaces to enable listening in the caching layer.
+
+```go
+type MultiStore interface {
+	...
+
+	// ListeningEnabled returns if listening is enabled for the KVStore belonging the provided StoreKey
+	ListeningEnabled(key StoreKey) bool
+
+	// SetListeners sets the WriteListeners for the KVStore belonging to the provided StoreKey
+	// It appends the listeners to a current set, if one already exists
+	SetListeners(key StoreKey, listeners []WriteListener)
+}
+```
+
+```go
+type CacheWrap interface {
+	...
+
+	// CacheWrapWithListeners recursively wraps again with listening enabled
+	CacheWrapWithListeners(storeKey types.StoreKey, listeners []WriteListener) CacheWrap
+}
+
+type CacheWrapper interface {
+	...
+
+	// CacheWrapWithListeners recursively wraps again with listening enabled
+	CacheWrapWithListeners(storeKey types.StoreKey, listeners []WriteListener) CacheWrap
+}
+```
+
+### MultiStore implementation updates
+
+We will modify all of the `Store` and `MultiStore` implementations to satisfy these new interfaces, and adjust the `rootmulti` `GetKVStore` method
+to wrap the returned `KVStore` with a `listenkv.Store` if listening is turned on for that `Store`.
+
+```go
+func (rs *Store) GetKVStore(key types.StoreKey) types.KVStore {
+	store := rs.stores[key].(types.KVStore)
+
+	if rs.TracingEnabled() {
+		store = tracekv.NewStore(store, rs.traceWriter, rs.traceContext)
+	}
+	if rs.ListeningEnabled(key) {
+		store = listenkv.NewStore(key, store, rs.listeners[key])
+	}
+
+	return store
+}
+```
+
+We will also adjust the `cachemulti` constructor methods and the `rootmulti` `CacheMultiStore` method to forward the listeners
+to and enable listening in the cache layer.
+
+```go
+func (rs *Store) CacheMultiStore() types.CacheMultiStore {
+	stores := make(map[types.StoreKey]types.CacheWrapper)
+	for k, v := range rs.stores {
+		stores[k] = v
+	}
+	return cachemulti.NewStore(rs.db, stores, rs.keysByName, rs.traceWriter, rs.traceContext, rs.listeners)
+}
+```
+
+### Exposing the data
+
+We will introduce a new `StreamingService` interface for exposing `WriteListener` data streams to external consumers.
+
+```go
+// Hook interface used to hook into the ABCI message processing of the BaseApp
+type Hook interface {
```

Comment:
Very nice.
How are these messages interlaced to with the `WriteListener` messages?
To easily see which writes occured in EndBlock, which in a given tx, etc.

Reply:
(answered below)